File: iOS/Mobile/DirectoryViewController.m
//
//  DirectoryViewController.m
//  Mobile
//
//  Created by Jason Hocker on 10/4/12.
//  Copyright (c) 2012 Ellucian. All rights reserved.
//

#import <AddressBook/AddressBook.h>
#import "DirectoryViewController.h"
#import "DirectoryEntry.h"
#import "CurrentUser.h"
#import "Base64.h"
#import "AppDelegate.h"
#import "DirectoryEntryViewController.h"
#import "UIViewController+GoogleAnalyticsTrackerSupport.h"
#import "AppearanceChanger.h"

@interface DirectoryViewController ()

@property (nonatomic, strong) NSURLConnection *connection;
@property (nonatomic, strong) NSMutableData *data;
@property (nonatomic, assign) ABPersonSortOrdering sort;
@property (nonatomic, assign) SEL sortSelector;
@property (nonatomic, assign) UILocalizedIndexedCollation *collation;
@property (readwrite, copy, nonatomic) NSArray *tableData;
@property (nonatomic, assign) BOOL firstNameFirst;
@property (nonatomic, strong) NSString *selectedScope;
@end

@implementation DirectoryViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    ABAddressBookRef addressBook;
    addressBook = ABAddressBookCreateWithOptions(NULL, NULL);
    __block BOOL accessGranted = NO;
    if (ABAddressBookRequestAccessWithCompletion != NULL) { // we're on iOS 6
        dispatch_semaphore_t sema = dispatch_semaphore_create(0);
        ABAddressBookRequestAccessWithCompletion(addressBook, ^(bool granted, CFErrorRef error) {
            accessGranted = granted;
            dispatch_semaphore_signal(sema);
        });
        dispatch_semaphore_wait(sema, DISPATCH_TIME_FOREVER);
        //dispatch_release(sema);
    }
    else { // we're on iOS 5 or older
        accessGranted = YES;
    }
    
    if(accessGranted) {
        self.sort = ABPersonGetSortOrdering();
        self.sortSelector = self.sort == kABPersonCompositeNameFormatFirstNameFirst ? @selector(firstName) : @selector(lastName);
    
        self.firstNameFirst = ABPersonGetCompositeNameFormat() == kABPersonCompositeNameFormatFirstNameFirst;
        CFRelease(addressBook);
    } else {
        self.sort = kABPersonCompositeNameFormatFirstNameFirst;
        self.sortSelector = @selector(lastName);
        self.firstNameFirst = YES;
    }


    self.title = self.module.name;
    self.tableData = [self partitionObjects:[[NSArray alloc] init] collationStringSelector:self.sortSelector];
    
    if(self.initialScope) {
        self.selectedScope = self.initialScope;
    }
    NSMutableArray *scopes = [[NSMutableArray alloc] init];
    if(self.module && [self.module.type isEqualToString:@"directory"]) {
        if([[self.module propertyForKey:@"student"] isEqualToString:@"true"] && !self.hideStudents) {
            if(!self.selectedScope) self.selectedScope = kDirectoryViewTypeStudent;
            [scopes addObject:kDirectoryViewTypeStudent ];
        }
        if([[self.module propertyForKey:@"faculty"] isEqualToString:@"true"] && !self.hideFaculty) {
            if(!self.selectedScope) self.selectedScope = kDirectoryViewTypeFaculty;
            [scopes addObject:kDirectoryViewTypeFaculty ];
        }
        if([scopes count] > 1) [scopes addObject:kDirectoryViewTypeAll ];
    } else {
        if([[NSUserDefaults standardUserDefaults] objectForKey:@"urls-directory-studentSearch"] && !self.hideStudents) {
            if(!self.selectedScope) self.selectedScope = kDirectoryViewTypeStudent;
            [scopes addObject:kDirectoryViewTypeStudent ];
        }
        if([[NSUserDefaults standardUserDefaults] objectForKey:@"urls-directory-facultySearch"] && !self.hideFaculty ) {
            if(!self.selectedScope) self.selectedScope = kDirectoryViewTypeFaculty;
            [scopes addObject:kDirectoryViewTypeFaculty ];
        }
        if([[NSUserDefaults standardUserDefaults] objectForKey:@"urls-directory-allSearch"] && [scopes count] > 1) {
            if(!self.selectedScope) self.selectedScope = kDirectoryViewTypeAll;
            [scopes addObject:kDirectoryViewTypeAll ];
        }
    }
    
    if([scopes count] > 1) {
        self.searchBar.scopeButtonTitles = [scopes copy];

        if(SYSTEM_VERSION_LESS_THAN(@"7.0")) {
            self.searchBar.tintColor = [UIColor primaryColor];
        }

        if(self.initialScope) {
            self.searchBar.selectedScopeButtonIndex = [self.searchBar.scopeButtonTitles indexOfObject:self.initialScope];
        } else {
            self.searchBar.selectedScopeButtonIndex = 0;
        }
        self.searchBar.showsScopeBar = YES;
        [self.searchBar sizeToFit];
    } else {
        self.searchBar.scopeButtonTitles = nil;
        self.searchBar.showsScopeBar = NO;
        [self.searchBar sizeToFit];
    }
  
    if(self.initialQueryString) {
        self.searchBar.text = self.initialQueryString;
        [self search:self.initialQueryString];
    }
}


- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    
    [self sendView:@"Directory page" forModuleNamed:self.module.name];
    
    if(!self.initialQueryString) {
        [self.searchBar becomeFirstResponder];
    }
}

- (void)searchBarCancelButtonClicked:(UISearchBar *)searchBar
{
    [self.searchBar resignFirstResponder];
    [self clear];
}

- (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText
{
    [self search:searchBar.text];
}

- (void)searchBarSearchButtonClicked:(UISearchBar *)searchBar
{
    [searchBar resignFirstResponder];
}

- (void)connection:(NSURLConnection *)theConnection didReceiveData:(NSData *)incrementalData {
    if (self.data==nil) {
        self.data = [[NSMutableData alloc] initWithCapacity:2048];
    }
    [self.data appendData:incrementalData];
    
    
    [UIApplication sharedApplication].networkActivityIndicatorVisible = YES;
}

- (void)connectionDidFinishLoading:(NSURLConnection*)theConnection {
    [self parseResponse:self.data];
    [UIApplication sharedApplication].networkActivityIndicatorVisible = NO;
    self.data = nil;
    self.connection = nil;

}

- (void) parseResponse:(NSData *)responseData
{
    NSMutableArray *tempContacts = [[NSMutableArray alloc] init];
    NSError *error;
    if(responseData)
    {
        
        NSDictionary* json = [NSJSONSerialization
                              JSONObjectWithData:responseData
                              options:kNilOptions
                              error:&error];
        
        //create/update objects
        for(NSDictionary *entry in [json objectForKey:@"entries"]) {
            DirectoryEntry *directoryEntry = [[DirectoryEntry alloc] init];

            if([entry objectForKey:@"personId"] != [NSNull null]) {
                directoryEntry.personId = [entry objectForKey:@"personId"];
            }
            if([entry objectForKey:@"username"] != [NSNull null]) {
                directoryEntry.username = [entry objectForKey:@"username"];
            }
            if([entry objectForKey:@"displayName"] != [NSNull null]) {
                directoryEntry.displayName = [entry objectForKey:@"displayName"];
            }
            if([entry objectForKey:@"firstName"] != [NSNull null]) {
                directoryEntry.firstName = [entry objectForKey:@"firstName"];
            }
            if([entry objectForKey:@"middleName"] != [NSNull null]) {
                directoryEntry.middleName = [entry objectForKey:@"middleName"];
            }
            if([entry objectForKey:@"lastName"] != [NSNull null]) {
                directoryEntry.lastName = [entry objectForKey:@"lastName"];
            }
            if([entry objectForKey:@"title"] != [NSNull null]) {
                directoryEntry.title = [entry objectForKey:@"title"];
            }
            if([entry objectForKey:@"office"] != [NSNull null]) {
                directoryEntry.office = [entry objectForKey:@"office"];
            }
            if([entry objectForKey:@"department"] != [NSNull null]) {
                directoryEntry.department = [entry objectForKey:@"department"];
            }
            if([entry objectForKey:@"phone"] != [NSNull null]) {
                directoryEntry.phone = [entry objectForKey:@"phone"];
            }
            if([entry objectForKey:@"mobile"] != [NSNull null]) {
                directoryEntry.mobile = [entry objectForKey:@"mobile"];
            }
            if([entry objectForKey:@"email"] != [NSNull null]) {
                directoryEntry.email = [entry objectForKey:@"email"];
            }
            if([entry objectForKey:@"street"] != [NSNull null]) {
                directoryEntry.street = [[entry objectForKey:@"street"] stringByReplacingOccurrencesOfString:@"\\n" withString:@"\n"];
            }
            if([entry objectForKey:@"room"] != [NSNull null]) {
                directoryEntry.room = [entry objectForKey:@"room"];
            }
            if([entry objectForKey:@"postOfficeBox"] != [NSNull null]) {
                directoryEntry.postOfficeBox = [entry objectForKey:@"postOfficeBox"];
            }
            if([entry objectForKey:@"city"] != [NSNull null]) {
                directoryEntry.city = [entry objectForKey:@"city"];
            }
            if([entry objectForKey:@"state"] != [NSNull null]) {
                directoryEntry.state = [entry objectForKey:@"state"];
            }
            if([entry objectForKey:@"postalCode"] != [NSNull null]) {
                directoryEntry.postalCode = [entry objectForKey:@"postalCode"];
            }
            if([entry objectForKey:@"country"] != [NSNull null]) {
                directoryEntry.country = [entry objectForKey:@"country"];
            }
            if([entry objectForKey:@"prefix"] != [NSNull null]) {
                directoryEntry.prefix = [entry objectForKey:@"prefix"];
            }
            if([entry objectForKey:@"suffix"] != [NSNull null]) {
                directoryEntry.suffix = [entry objectForKey:@"suffix"];
            }
            [tempContacts addObject:directoryEntry];
        }
    }
    
    if(self.sort == kABPersonSortByFirstName) {
        [tempContacts sortUsingDescriptors:[NSArray arrayWithObject: [NSSortDescriptor sortDescriptorWithKey:@"firstName" ascending:YES]]];
    } else {
        [tempContacts sortUsingDescriptors:[NSArray arrayWithObject: [NSSortDescriptor sortDescriptorWithKey:@"lastName" ascending:YES]]];
    }

    self.tableData = [self partitionObjects:[tempContacts copy] collationStringSelector:self.sortSelector];
    [self.tableView reloadData];
}



- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    DirectoryEntry *entry = nil;
    entry = [[self.tableData objectAtIndex:indexPath.section] objectAtIndex:indexPath.row];

    static NSString *CellIdentifier = @"Directory Name Cell";

    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier];
        cell.accessoryType = UITableViewCellAccessoryNone;
    }

    UILabel *textLabel = (UILabel *)[cell viewWithTag:1];
    textLabel.text = [entry nameOrderedByFirstName:self.firstNameFirst];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    DirectoryEntry *entry = [[self.tableData objectAtIndex:indexPath.section] objectAtIndex:indexPath.row];
    [self performSegueWithIdentifier:@"Show Directory Profile" sender:entry];
}

- (void)searchBar:(UISearchBar *)searchBar selectedScopeButtonIndexDidChange:(NSInteger)selectedScope
{
    [self sendEventWithCategory:kAnalyticsCategoryUI_Action withAction:kAnalyticsActionButton_Press withLabel:@"Select directory type" withValue:nil forModuleNamed:self.module.name];
    self.selectedScope = [self.searchBar.scopeButtonTitles objectAtIndex:self.searchBar.selectedScopeButtonIndex];
    if(searchBar.text.length > 0) {
        [self search:searchBar.text];
    }
}

-(void)search:(NSString *)searchString
{
    self.initialQueryString = searchString;
    if([searchString length] > 0) {
        NSString *url = nil;
        NSString *scopeTitle = self.selectedScope;
        if(self.module && [self.module.type isEqualToString:@"directory"]) {
            if([scopeTitle isEqualToString:kDirectoryViewTypeStudent]) {
                url = [self.module propertyForKey:@"studentSearch"];
            } else if([scopeTitle isEqualToString:kDirectoryViewTypeFaculty]) {
                url = [self.module propertyForKey:@"facultySearch"];
            } else {
                 url = [self.module propertyForKey:@"allSearch"];
            }
        } else {
            if([scopeTitle isEqualToString:kDirectoryViewTypeStudent]) {
                url = [[NSUserDefaults standardUserDefaults] objectForKey:@"urls-directory-studentSearch"];
            } else if([scopeTitle isEqualToString:kDirectoryViewTypeFaculty]) {
               url = [[NSUserDefaults standardUserDefaults] objectForKey:@"urls-directory-facultySearch"];
            } else {
               url = [[NSUserDefaults standardUserDefaults] objectForKey:@"urls-directory-allSearch"];
            }
        }
        url = [NSString stringWithFormat:@"%@?searchString=%@", url, [searchString stringByAddingPercentEscapesUsingEncoding: NSUTF8StringEncoding]];
        
        AppDelegate *appDelegate = (AppDelegate *)[[UIApplication sharedApplication] delegate];
        CurrentUser *user = [appDelegate currentUser];
        NSMutableString *loginString = (NSMutableString*)[@"" stringByAppendingFormat:@"%@:%@", user.userauth, user.getPassword];

        NSString *encodedLoginData = [Base64 encode:[loginString dataUsingEncoding:NSUTF8StringEncoding]];

        NSString *authHeader = [@"Basic " stringByAppendingFormat:@"%@", encodedLoginData];
        
        NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:url] cachePolicy:NSURLRequestUseProtocolCachePolicy timeoutInterval:30.0];
        [request setValue:authHeader forHTTPHeaderField:@"Authorization"];

        NSLog(@"%@",url);
        if(self.connection!=nil){ //cancel if in process
            [self.connection cancel];
        }
        self.connection = [[NSURLConnection alloc] initWithRequest:request delegate:self];
    } else {
        [self clear];
    }
}

-(NSArray *)partitionObjects:(NSArray *)array collationStringSelector:(SEL)selector

{
    UILocalizedIndexedCollation *collation = [UILocalizedIndexedCollation currentCollation];
    
    NSInteger sectionCount = [[collation sectionTitles] count]; //section count is take from sectionTitles and not sectionIndexTitles
    NSMutableArray *unsortedSections = [NSMutableArray arrayWithCapacity:sectionCount];
    
    //create an array to hold the data for each section
    for(int i = 0; i < sectionCount; i++)
    {
        [unsortedSections addObject:[NSMutableArray array]];
    }
    
    //put each object into a section
    for (id object in array)
    {
        NSInteger index = [collation sectionForObject:object collationStringSelector:selector];
        [[unsortedSections objectAtIndex:index] addObject:object];
    }
    
    NSMutableArray *sections = [NSMutableArray arrayWithCapacity:sectionCount];
    
    //sort each section
    for (NSMutableArray *section in unsortedSections)
    {
        [sections addObject:[collation sortedArrayFromArray:section collationStringSelector:selector]];
    }
    
    return sections;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [[self.tableData objectAtIndex:section] count];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return [[[UILocalizedIndexedCollation currentCollation] sectionTitles] count];
}

- (NSString *)tableView:(UITableView *)tableView stringForTitleForHeaderInSection:(NSInteger)section
{
    BOOL showSection = [[self.tableData objectAtIndex:section] count] != 0;
    //only show the section title if there are rows in the section
    return (showSection) ? [[[UILocalizedIndexedCollation currentCollation] sectionTitles] objectAtIndex:section] : nil;
}

- (NSInteger)tableView:(UITableView *)tableView sectionForSectionIndexTitle:(NSString *)title atIndex:(NSInteger)index
{
    return [[UILocalizedIndexedCollation currentCollation] sectionForSectionIndexTitleAtIndex:index];
}

-(void) clear
{
    self.tableData = [self partitionObjects:[[NSArray alloc] init] collationStringSelector:self.sortSelector];
    [self.tableView reloadData];

}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    
    if ([[segue identifier] isEqualToString:@"Show Directory Profile"])
    {
        DirectoryEntryViewController *vc = (DirectoryEntryViewController *)[segue destinationViewController];
        vc.entry = sender;
        vc.module = self.module;
    }
}

@end

File: iOS/Mobile/DirectoryViewController.swift
//
//  DirectoryViewController.swift
//  Mobile
//
//  Created by Jason Hocker on 12/2/15.
//  Copyright © 2015 Ellucian Company L.P. and its affiliates. All rights reserved.
//

import Foundation
import Contacts

class DirectoryViewController : UITableViewController, UISearchBarDelegate, DirectoryFilterDelegate, EllucianMobileLaunchableControllerProtocol {
    
    let searchDelayInterval : TimeInterval = 0.5
    
    @IBOutlet var searchBar: UISearchBar!
    var module : Module!
    var hideStudents : Bool?
    var hideFaculty : Bool?
    
    var forcedFilteredGroup : String?
    var searchDelayer : Timer?
    var legacySearch = false
    var hiddenGroups = [String]()
    
    @IBOutlet var filterButton: UIBarButtonItem!
    var tableData = [[DirectoryEntry]]()
    var entries = [DirectoryEntry]()
    var data : NSMutableData?
    var groups = [DirectoryDefinitionProtocol]()
    var inSearch = false
    
    let defaultSession = URLSession(configuration: .default)
    var dataTask : URLSessionDataTask?
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        if entries.count > 0 {
            //coming from another source, such as course roster
            inSearch = true
            buildTableData(self.entries)
        }
        
        buildGroups()
        
        tableView.rowHeight = UITableViewAutomaticDimension
        tableView.estimatedRowHeight = 53
        
        self.title = self.module?.name
        
        if UIScreen.main.traitCollection.userInterfaceIdiom == .pad {
            self.splitViewController?.preferredDisplayMode = .allVisible;
        } else {
            self.splitViewController?.preferredDisplayMode = .automatic;
        }
        
        if groups.count <= 1 {
            filterButton.isEnabled = false;
            self.navigationController?.navigationBar.topItem!.rightBarButtonItem = nil;
        }
        
        if let searchBar = searchBar {
            searchBar.becomeFirstResponder()
        }
        
        
        NotificationCenter.default.addObserver(self, selector: #selector(DirectoryViewController.signInHappened), name: CurrentUser.LoginExecutorSuccessNotification, object: nil)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        sendView("Directory page", moduleName: self.module?.name)
        
    }
    
    func doDelayedSearch(_ timer: Timer) {
        searchDelayer = nil
        doSearch()
    }
    
    func doSearch() {
        MBProgressHUD.hide(for: self.view, animated: true)
        let hud = MBProgressHUD.showAdded(to: self.view, animated: true)
        let loadingString = NSLocalizedString("Loading", comment: "loading message while waiting for data to load")
        hud.label.text = loadingString
        UIAccessibilityPostNotification(UIAccessibilityAnnouncementNotification, loadingString)
        
        if !self.legacySearch {
            doModernSearch()
        } else {
            doLegacySearch()
        }
    }
    
    func doLegacySearch() {
        if let searchString = self.searchBar.text {
            if searchString.characters.count > 0 {
                var url: String? = nil
                if let module = module , self.module?.type == "directory" {
                    
                    let groupsInUse = self.groups.filter() {
                        !self.hiddenGroups.contains($0.internalName!)
                    }
                    let filteredGroups = groupsInUse.map() {
                        $0.internalName!
                    }
                    
                    if filteredGroups.count == 0 {
                        clear()
                        return
                    } else if filteredGroups.count == 2 {
                        url = module.property(forKey: "allSearch")
                    } else if filteredGroups.count == 1 && filteredGroups.contains("student") {
                        url = module.property(forKey: "studentSearch")
                    } else if filteredGroups.count == 1 && filteredGroups.contains("faculty") {
                        url = module.property(forKey: "facultySearch")
                    }
                }
                
                let encodedSearchString = searchString.addingPercentEncoding(withAllowedCharacters: CharacterSet.urlQueryAllowed)
                url = "\(url!)?searchString=\(encodedSearchString!)"
                
                var request = URLRequest(url: URL(string: url!)!, cachePolicy: .useProtocolCachePolicy, timeoutInterval: 30.0)
                if (LoginExecutor.isUsingBasicAuthentication()) {
                    request.addAuthenticationHeader()
                }
                executeSearch(request)
            }
        } else {
            clear()
        }
    }
    
    func doModernSearch() {
        if let searchString = self.searchBar.text {
            if searchString.characters.count > 0 {
                var url = module?.property(forKey: "baseSearch")
                
                var groupsInUse = self.groups
                if self.hiddenGroups.count > 0 {
                    groupsInUse = self.groups.filter() {
                        !self.hiddenGroups.contains($0.internalName!)
                    }
                }
                
                let filteredGroups = groupsInUse.map() {
                    return $0.internalName as String!
                    } as [String]
                
                if filteredGroups.count > 0 {
                    
                    let encodedSearchString = searchString.addingPercentEncoding(withAllowedCharacters: CharacterSet.urlQueryAllowed)
                    let encodedDirectories = filteredGroups.joined(separator: ",").addingPercentEncoding(withAllowedCharacters: CharacterSet.urlQueryAllowed)
                    url = "\(url!)?searchString=\(encodedSearchString!)&directories=\(encodedDirectories!)"
                    var request = URLRequest(url: URL(string: url!)!, cachePolicy: .useProtocolCachePolicy, timeoutInterval: 30.0)
                    if CurrentUser.sharedInstance.isLoggedIn {
                        request.addAuthenticationHeader()
                    }
                    executeSearch(request)
                } else {
                    clear()
                }
            }
        } else {
            clear()
        }
    }
    
    func clear() {
        inSearch = false
        self.entries = [DirectoryEntry]()
        if !self.splitViewController!.isCollapsed {
            self.performSegue(withIdentifier: "Empty Entry", sender: nil)
        }
        self.tableData = self.partitionObjects(self.entries, collationStringSelector: #selector(DirectoryEntry.nameToUseForLastNameSort))
        self.tableView.reloadData()
    }
    
    // MARK: UISearchBarDelegate
    
    func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
        if let searchDelayer = searchDelayer {
            searchDelayer.invalidate()
        }
        if searchText.characters.count > 0 {
            searchDelayer = Timer.scheduledTimer(timeInterval: searchDelayInterval, target: self, selector: #selector(DirectoryViewController.doDelayedSearch(_:)), userInfo: searchText, repeats: false)
            inSearch = true
        } else {
            clear()
        }
    }
    
    func searchBarSearchButtonClicked(_ searchBar: UISearchBar) {
        searchBar.resignFirstResponder()
        searchDelayer = nil
        doSearch()
    }
    
    //MARK: segue
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "Show Filter" {
            self.sendEvent(category: .ui_Action, action: .button_Press, label: "Select directory type", moduleName: self.module?.name)
            let navigationController = segue.destination as! UINavigationController
            let detailController = navigationController.viewControllers[0] as! DirectoryFilterViewController
            detailController.module = self.module
            detailController.hiddenGroups = self.hiddenGroups
            detailController.groups = self.groups
            detailController.delegate = self
        } else if segue.identifier == "Show Directory Profile" {
            let detailController : DirectoryEntryViewController
            if segue.destination is DirectoryEntryViewController {
                detailController = segue.destination as! DirectoryEntryViewController
            } else {
                let navigationController = segue.destination as! UINavigationController
                detailController = navigationController.viewControllers[0] as! DirectoryEntryViewController
            }
            detailController.entry = sender as? DirectoryEntry;
            detailController.module = self.module;
        }
    }
    
    //MARK: other
    func partitionObjects(_ array: [DirectoryEntry], collationStringSelector selector: Selector) -> [[DirectoryEntry]] {
        let collation: UILocalizedIndexedCollation = UILocalizedIndexedCollation.current()
        let sectionCount = collation.sectionTitles.count
        //section count is take from sectionTitles and not sectionIndexTitles
        var unsortedSections = [[DirectoryEntry]](repeating: [DirectoryEntry](), count: sectionCount)
        
        //put each object into a section
        for object in array {
            let index = collation.section(for: object, collationStringSelector: selector)
            unsortedSections[index].append(object)
        }
        var sections = [[DirectoryEntry]](repeating: [DirectoryEntry](), count: sectionCount)
        //sort each section
        for i in 0 ..< unsortedSections.count {
            let sorted = collation.sortedArray(from: unsortedSections[i], collationStringSelector: selector) as! [DirectoryEntry]
            sections[i] = sorted
        }
        
        return sections;
    }
    
    func connection(_ theConnection: NSURLConnection, didReceiveData incrementalData: Data) {
        if self.data == nil {
            self.data = NSMutableData(capacity: 2048)
        }
        self.data?.append(incrementalData)
        
    }
    
    func buildTableData(_ entries: [DirectoryEntry]) {
        
        switch CNContactsUserDefaults.shared().sortOrder {
        case .none, .userDefault, .familyName:
            self.tableData = self.partitionObjects(self.entries, collationStringSelector: #selector(DirectoryEntry.nameToUseForLastNameSort))
        case .givenName:
            self.tableData = self.partitionObjects(self.entries, collationStringSelector: #selector(DirectoryEntry.nameToUseForFirstNameSort))
        }
        
        DispatchQueue.main.async(execute: {
            self.tableView.reloadData()
        })
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if section == 0 {
            var count = 0
            if inSearch && (entries.count == 0 || searchBar?.text?.characters.count == 0) {
                count += 1
            }
            if !inSearch {
                count += 1
            }
            if !CurrentUser.sharedInstance.isLoggedIn {
                count += 1
            }
            return count
        }
        if entries.count > 0 {
            return tableData[section-1].count
        }
        return 0
    }
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return UILocalizedIndexedCollation.current().sectionTitles.count + 1
    }
    
    var badImages = [String]()
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if (indexPath as NSIndexPath).section == 0 {
            switch ((indexPath as NSIndexPath).row, CurrentUser.sharedInstance.isLoggedIn, inSearch, entries.count == 0) {
            case (0, true, false, _), (0, false, false, _):
                return tableView.dequeueReusableCell(withIdentifier: "Directory Initial Message Cell", for: indexPath) as UITableViewCell
            case (0, false, true, _), (1, false, false, _):
                return tableView.dequeueReusableCell(withIdentifier: "Directory Sign In Message Cell", for: indexPath) as UITableViewCell
            case (0, true, true, true), (1, false, true, true):
                return tableView.dequeueReusableCell(withIdentifier: "Directory Empty Message Cell", for: indexPath) as UITableViewCell
            default:
                return UITableViewCell()
            }
        } else if (indexPath.section-1) < self.tableData.count && indexPath.row < self.tableData[indexPath.section-1].count {
            let entry = self.tableData[(indexPath as NSIndexPath).section-1][(indexPath as NSIndexPath).row]
            let cell : UITableViewCell
            if let logo = entry.imageUrl , logo != "" {
                if badImages.contains(logo) {
                    cell  = tableView.dequeueReusableCell(withIdentifier: "Directory Name Cell", for: indexPath) as UITableViewCell
                } else {
                    cell  = tableView.dequeueReusableCell(withIdentifier: "Directory Name Image Cell", for: indexPath) as UITableViewCell
                    let imageView = cell.viewWithTag(3) as! UIImageView
                    imageView.loadImagefromURL(logo, successHandler: {
                        imageView.isHidden = false
                        imageView.convertToCircleImage() }, failureHandler:  {
                            DispatchQueue.main.async {
                                () -> Void in
                                //still the image in the cell we care about?
                                let desiredSection = (indexPath).section-1
                                if desiredSection < self.tableData.count {
                                    let sectionData = self.tableData[desiredSection]
                                    let desiredRow = (indexPath as NSIndexPath).row
                                    
                                    if desiredRow < sectionData.count {
                                        let currentEntry = sectionData[desiredRow]
                                        if entry == currentEntry {
                                            
                                            imageView.isHidden = true
                                            self.badImages.append(logo)
                                            self.tableView.reloadRows(at: [indexPath], with: .none)
                                        }
                                    }
                                }
                            }
                            
                            
                        }
                    )
                }
            } else {
                cell  = tableView.dequeueReusableCell(withIdentifier: "Directory Name Cell", for: indexPath) as UITableViewCell
            }
            
            let nameLabel = cell.viewWithTag(1) as! UILabel
            let typeLabel = cell.viewWithTag(2) as! UILabel
            nameLabel.text = entry.nameToUseForDisplay()
            typeLabel.text = formatType(entry)
            return cell
        }
        return UITableViewCell()
    }
    
    func formatType(_ entry: DirectoryEntry) -> String? {
        if let type = entry.type {
            if !legacySearch {
                return type
            } else {
                switch type {
                case "student": return NSLocalizedString("Students", comment: "student search scope in directory")
                case "faculty": return NSLocalizedString("Faculty/Staff", comment:"facilty/staff search scope in directory")
                default: return type
                    
                }
            }
            
        }
        return nil
    }
    
    override func tableView(_ tableView: UITableView, willSelectRowAt indexPath: IndexPath) -> IndexPath? {
        return (indexPath as NSIndexPath).section == 0 ? nil : indexPath
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let entry = self.tableData[(indexPath as NSIndexPath).section-1][(indexPath as NSIndexPath).row]
        self.performSegue(withIdentifier: "Show Directory Profile", sender: entry)
    }
    
    func updateFilter(_ filter: [String]) {
        if !self.splitViewController!.isCollapsed {
            self.performSegue(withIdentifier: "Empty Entry", sender: nil)
        }
        self.hiddenGroups = filter
        if groups.count == filter.count {
            clear()
        } else if let searchString = self.searchBar.text, searchString.characters.count > 0 {
            doSearch()
        }
    }
    
    func signInHappened() {
        DispatchQueue.main.async(execute: {
            self.tableView.reloadData()
        })
    }
    
    func buildGroups () {
        
        
        let moduleKey = "\(module!.internalKey!)-hiddenGroups"
        if let hiddenGroups = AppGroupUtilities.userDefaults()?.array(forKey: moduleKey) as? [String] {
            self.hiddenGroups = hiddenGroups
        } else {
            self.hiddenGroups = [String]()
        }
        
        do {
            let fetchRequest = NSFetchRequest<DirectoryDefinition>(entityName: "DirectoryDefinition")
            fetchRequest.sortDescriptors = [ NSSortDescriptor(key: "displayName", ascending: true)]
            
            if let groups = try self.module!.managedObjectContext?.fetch(fetchRequest) , groups.count > 0 && module?.property(forKey: "directories") != nil {
                let supportedDirectories = module?.property(forKey: "directories")?.components(separatedBy: ",")
                let moduleGroups = groups.filter() {
                    supportedDirectories!.contains($0.key!)
                }
                self.groups = moduleGroups
            } else {
                legacySearch = true
                if self.module?.property(forKey: "student") == "true" && self.module?.property(forKey: "faculty") == "true" {
                    self.groups = [ legacyStudentDefinition(), legacyFacultyDefinition() ]
                } else if self.module?.property(forKey: "student") == "true" {
                    self.groups = [ legacyStudentDefinition() ]
                } else if self.module?.property(forKey: "faculty") == "true" {
                    self.groups = [ legacyFacultyDefinition() ]
                } else {
                    self.groups = [ legacyStudentDefinition(), legacyFacultyDefinition() ]
                }
                self.groups = self.groups.sorted( by: { $0.displayName! < $1.displayName!})
            }
        } catch {
        }
    }
    
    func legacyStudentDefinition() -> DirectoryDefinitionProtocol {
        let dir = DirectoryUnmanagedDefinition()
        dir.internalName = "student"
        dir.displayName = NSLocalizedString("Students", comment: "student search scope in directory")
        dir.authenticatedOnly = true
        return dir
    }
    
    func legacyFacultyDefinition() -> DirectoryDefinitionProtocol {
        let dir = DirectoryUnmanagedDefinition()
        dir.internalName = "faculty"
        dir.displayName = NSLocalizedString("Faculty/Staff", comment:"facilty/staff search scope in directory")
        dir.authenticatedOnly = true
        return dir
    }
    
    private func executeSearch(_ request: URLRequest) {
        dataTask = defaultSession.dataTask(with: request) {
            data, response, error in
            if self.inSearch {
                if let data = data {
                    self.entries = DirectoryEntry.parseResponse(data)
                    self.buildTableData(self.entries)
                } else {
                    DispatchQueue.main.async {
                        let alertController = UIAlertController(title: NSLocalizedString("Poor Network Connection", comment:"title when data cannot load due to a poor netwrok connection"), message: NSLocalizedString("Data could not be retrieved.", comment:"message when data cannot load due to a poor netwrok connection"), preferredStyle: .alert)
                        let alertAction = UIAlertAction(title: NSLocalizedString("OK", comment:"OK"), style: UIAlertActionStyle.default)
                        alertController.addAction(alertAction)
                        self.present(alertController, animated: true)
                    }
                }
            }
            UIApplication.shared.isNetworkActivityIndicatorVisible = false
            DispatchQueue.main.async(execute: {
                MBProgressHUD.hide(for: self.view, animated: true)
            })
        }
        UIApplication.shared.isNetworkActivityIndicatorVisible = true
        dataTask?.resume()
    }
}
